src/Hosting/Systemd/src/SystemdHelpers.cs

```diff
@@ -12,14 +16,55 @@ public static class SystemdHelpers
     {
         private const string INVOCATION_ID = "INVOCATION_ID";
 
+        private static bool? _isSystemdService;
+
         /// <summary>
         /// Check if the current process is hosted as a systemd Service.
         /// </summary>
         /// <returns><c>True</c> if the current process is hosted as a systemd Service, otherwise <c>false</c>.</returns>
         public static bool IsSystemdService()
+            => _isSystemdService ?? (bool)(_isSystemdService = CheckSystemdUnit());
+
+        private static bool CheckSystemdUnit()
         {
-            // We use the INVOCATION_ID envirionment variable that was introduced in systemd 232 (released 2016-11-03).
-            return !string.IsNullOrEmpty(Environment.GetEnvironmentVariable(INVOCATION_ID));
+            // No point in testing anything unless it's Unix
+            if (Environment.OSVersion.Platform != PlatformID.Unix)
+            {
+                return false;
+            }
+
+            // We've got invocation id, it's systemd >= 232 running a unit
+            if (!string.IsNullOrEmpty(Environment.GetEnvironmentVariable(INVOCATION_ID)))
+            {
+                return true;
+            }
+
+            // Either it's not a unit, or systemd is < 232, do a bit more digging
+            try
+            {
+                // Test parent process
+                var parentPid = GetParentPid();
+                var ppidString = parentPid.ToString(NumberFormatInfo.InvariantInfo);
+
+                // If parent PID is not 1, this may be a user unit, in this case it must match MANAGERPID envvar
+                if (parentPid != 1
+                    && Environment.GetEnvironmentVariable("MANAGERPID") != ppidString)
+                {
+                    return false;
+                }
+
+                // Check parent process name to match "systemd\n"
+                var comm = File.ReadAllBytes("/proc/" + ppidString + "/comm");
+                return comm.AsSpan().SequenceEqual(Encoding.ASCII.GetBytes("systemd\n"));
+            }
+            catch
+            {
+            }
+
+            return false;
         }
+
+        [DllImport("libc", EntryPoint = "getppid", SetLastError = true)]
```

Comment:
I hadn't noticed before. This doesn't need `SetLastError`, `getppid` is always successful.

Reply:
Thanks for noticing this, pushed a fix.